README.md
RFLink
======

This is an Arduino library for managing level 2 ("link") communication with RF
devices.
Tested and used with Texas Instruments (tm) CC1101 device.
Works on top of arduino-CC1101 library (see below).

The link layer implements:
  - Packet ID, to properly ignore already-received packets
  - ACK, so that the sender will know data good reception

The notion of sender and receiver address is managed by RFLink, although this
is in part already managed as a built-in feature of CC1101.


Installation
------------

Download a zip of this repository, then include it from the Arduino IDE.

**IMPORTANT**
You'll also need to install arduino-cc1101 library, available here:
[https://github.com/veonik/arduino-cc1101](https://github.com/veonik/arduino-cc1101)


Usage
-----

See [examples/example1/sender/sender.ino](examples/example1/sender/sender.ino)
and
[examples/example1/receiver/receiver.ino](examples/example1/receiver/receiver.ino)
for an example.



ChangeLog
2020-03-03 Sébastien Millet <milletseb@laposte.net>

	* Fixes a few minor bugs in code
	* Fixes example links in README

2020-01-05 Sébastien Millet <milletseb@laposte.net>

	* Release 0.8 built


cc1101wrapper.h
// vim:ts=4:sw=4:tw=80:et
/*
  cc1101wrapper.h

  Header file of cc1101wrapper.cpp
*/

/*
  Copyright 2020 Sébastien Millet

  rflink is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  rflink is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this program. If not, see
  <https://www.gnu.org/licenses>.
*/

#ifndef _CC1101WRAPPER_H
#define _CC1101WRAPPER_H

#include "rflink.h"

#include <Arduino.h>

// Attach CC1101 pins to their corresponding SPI pins
// Uno or Nano pins:
//   CSN  (SS)   => 10
//   MOSI (SI)   => 11
//   MISO (SO)   => 12
//   SCK  (SCLK) => 13
//   GD0  (GDO0) => A valid interrupt pin for your platform (defined below this)
//                  Typically: 2
//   GD2  (GDO2) => In the set {2, 3}, the one not used by GDO0.
//                  Typically: 3
//
// * ************** **********************************************************
// * VERY IMPORTANT **********************************************************
// * ************** **********************************************************
// *
// * ON MANY SCHEMATICS FOUND ON INTERNET, GDO0 IS CONNECTED ON ARDUINO' D3.
// * THIS IS NOT COMPATIBLE WITH CC1101 DEFAULT LIB SETTINGS, THAT ASSUMES GDO0
// * IS CONNECTED ON D2.
#if defined(__AVR_ATmega2560__) || defined(__AVR_ATmega1280__)
#define CC1101Interrupt 4 // Pin 19
#define CC1101_GDO0 19
#elif defined(__MK64FX512__)
// Teensy 3.5
#define CC1101Interrupt 9 // Pin 9
#define CC1101_GDO0 9
#else
#define CC1101Interrupt 0 // Pin 2
#define CC1101_GDO0 2
#endif

void cc1101_attach(RFLink* link);

#endif // _CC1101WRAPPER_H



cc1101wrapper.cpp
// vim:ts=4:sw=4:tw=80:et
/*
  cc1101wrapper.cpp

  Intermediate code to make work together:
    arduino-cc1101 library published on GitHub under veonik, url:
      https://github.com/veonik/arduino-cc1101
  and
    rflink library

  NOTE

    This is NOT a 'driver' code for CC1101. The driver code is contained in
    arduino-cc1101 library (early 2020: cc1101.h, cc1101.cpp, ccpacket.h).

    The current file is just a quick 'helper' file, that aims to:
    - Avoid any CC1101-specific code inside rflink.{h, cpp}
    - Avoid (as much as possible) too much CC1101-specific code in the code that
      uses rflink
*/

/*
  Copyright 2020 Sébastien Millet

  rflink is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  rflink is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this program. If not, see
  <https://www.gnu.org/licenses>.
*/

#include "cc1101wrapper.h"
#include "rflink.h"

#include <cc1101.h>
#include <ccpacket.h>

//#define CC1101WRAPPER_DEBUG

#ifndef CC1101WRAPPER_DEBUG

#define dbg(a)
#define dbgf(...)
#define dbgbin(a, b, c)

#else

#include "debug.h"

#endif

CC1101 radio;
byte syncWord[2] = {0xA9, 0x5A};

void cc1101_init(byte* max_data_len, bool reset_only) {
    if (reset_only) {
        dbg("Resetting radio...");
        radio.reset();
        dbg("Radio reset done");
        return;
    }
    radio.init();
    radio.setSyncWord(syncWord);
    radio.setCarrierFreq(CFREQ_868);
    radio.enableAddressCheck();
    if (max_data_len)
        *max_data_len = (CCPACKET_DATA_LEN);
}

void cc1101_set_opt(opt_t opt, void *data, byte len) {
    if (opt == OPT_ADDRESS && len == 1) {
        // Set device address
        byte addr = *(byte*)data;
        radio.setDevAddress(addr);
        dbgf("Set device address to: 0x%02x", addr);

    } else if (opt == OPT_EMISSION_POWER && len == 1) {
        // Set power of device, when sending data
        // The data passed in must be a byte, where:
        //   a zero means low power
        //   any non zero means high power (a.k.a "long distance")
        byte level = *(byte*)data;
        byte pa_value;
        if (!level) {
            pa_value = PA_LowPower;
            dbg("Set device PA to low power");
        } else {
            pa_value = PA_LongDistance;
            dbg("Set device PA to high power");
        }
        radio.setTxPowerAmp(pa_value);

    } else if (opt == OPT_SNIF_MODE && len == 1) {
        byte val = *(byte*)data;
        if (val) {
            radio.disableAddressCheck();
            dbg("Disabled address check (a.k.a. snif mode)");
        } else {
            radio.enableAddressCheck();
            dbg("Enabled address check (a.k.a. non-snif mode)");
        }

    } else {
        dbgf("Error: unknown option code: %i", opt);
    }
}

// FIXME
// cc1101wrapper copies data from caller before handing over to
// sendData method of radio object.
// Using user data directly would be best.
byte cc1101_send(const void *data, byte len) {
    CCPACKET packet;
    packet.length = len;
    memcpy(packet.data, data, len);

    dbgf("cc1101_send: sending packet of %i byte(s):", len);
    dbgbin("cc1101_send:   ", packet.data, len);
    bool r = radio.sendData(packet);
    dbgf("cc1101_send: sendData return value: %i", r);

    return r ? ERR_OK : ERR_SEND_IO;
}

// FIXME
// Same remark as with cc1101_send: a lot of memcpy in the end, in the
// way it is designed today.
byte cc1101_receive(void *buf, byte buf_len) {
    CCPACKET packet;
    byte len;
    if ((len = radio.receiveData(&packet))) {

        dbgf("cc1101_receive: %i byte(s) packet received:", len);
        dbgbin("cc1101_receive:   ", packet.data, len);

        if (len > buf_len)
            len = buf_len;

        memcpy(buf, packet.data, len);
        return len;
    } else {
        return 0;
    }
}

void cc1101_set_interrupt(void (*func)()) {
    attachInterrupt(CC1101Interrupt, func, FALLING);
}

void cc1101_reset_interrupt() {
    detachInterrupt(CC1101Interrupt);
}

void cc1101_attach(RFLink* link) {
    RFLinkFunctions f;
    f.deviceInit = cc1101_init;
    f.deviceSend = cc1101_send;
    f.deviceReceive = cc1101_receive;
    f.deviceSetOpt = cc1101_set_opt;

    f.setInterrupt = cc1101_set_interrupt;
    f.resetInterrupt = cc1101_reset_interrupt;

    link->register_funcs(&f);
}



rflink.h
// vim:ts=4:sw=4:tw=80:et
/*
  rflink.h

  Header of rflink.cpp
*/

/*
  Copyright 2020 Sébastien Millet

  rflink is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  rflink is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this program. If not, see
  <https://www.gnu.org/licenses>.
*/

#ifndef _RFLINK_H
#define _RFLINK_H

//#define ERR_STRINGS

// If DEBUG is activated (said differently: if you uncomment RFLINK_DEBUG), then
// you got to include the debug library (debug.h, debug.cpp) in the compilation.
//#define RFLINK_DEBUG
//#define RFLINK_DEBUG_EVENTTIMER
//#define RFLINK_DEBUG_EVENTTIMER_ONLY

// Don't uncomment the below unless you know what you are doing...
//#define DEBUG_KEEP_SENDING_EVEN_AFTER_RECEIVING_ACK

#include <Arduino.h>

#define ENFORCE_MAX_TASK_COUNT_AT_COMPILE_TIME
#define DEFAULT_MAX_TASK_COUNT                15
#define DEFAULT_PRE_ALLOCATE                   0
#define PKTID_CACHE_SIZE                      10

// Delays below are in milliseconds
#define DEFAULT_RECEIVE_DATA_AVAIL_DELAY     900
#define DEFAULT_RECEIVE_PURGE_DELAY         1000
#define DEFAULT_RECEIVE_TIMEOUT_DELAY          0
#define DEFAULT_SEND_PURGE_DELAY            1000
// The below value makes 49 hours.
#define CACHE_PKTID_DISCARD_DELAY      176400000

#define MIN_DEVICE_RESET_DELAY              1000

#define POST_DEVICE_RESET_DELAY                1

#define ERR_OK                                 0
#define ERR_DEVICE_NOT_REGISTERED              1
#define ERR_SEND_FUNC_NOT_REGISTERED           2
#define ERR_RECEIVE_FUNC_NOT_REGISTERED        3
#define ERR_SEND_DATA_LEN_ABOVE_LIMIT          4
#define ERR_SEND_IO                            5
#define ERR_SEND_BAD_ARGUMENTS                 6
#define ERR_SEND_NO_ACK_RCVD                   7
#define ERR_TASK_CREATED_OK                    8
#define ERR_UNABLE_TO_CREATE_TASK              9
#define ERR_UNKNOWN_TASKID                    10
#define ERR_UNDEFINED                         11
#define ERR_TASK_UNDERWAY                     12
#define ERR_TIMEOUT                           13

// NOTE
// rflink.cpp assumes an address is 1-byte.
// If not, rflink.cpp code MUST be reviewed.
typedef uint8_t address_t;
static_assert(sizeof(address_t) == 1,
  "don't know how to handle not 1-byte addresses");
#define ASSUME_DEVICE_ADDRESS_IS_ONE_BYTE // rflink.cpp will not compile if this
                                          // define does not exist
#define ADDR_BROADCAST                      0xFF

typedef uint16_t pktid_t;

typedef uint16_t taskid_t;

// "m" like milliseconds
typedef long unsigned int mtime_t;

struct Header {
    /*
     *  WARNING
     *
     *  On CC1101, the first byte MUST be the destination address.
     */
    address_t dst;
    address_t src;
    uint8_t flags;
    pktid_t pktid;
    uint8_t len;
};

#define FLAG_NONE 0
#define FLAG_SIN  (1 << 0)
#define FLAG_ACK  (1 << 1)

struct Packet {
    Header header;
    // 'data' member is never used neither "malloc'ed".
    // Used to extract address of first byte of data AFTER packet header.
    //
    // This position can be referred to as:
    //   &pkt->header + 1
    // (therefore, without the need for the below definition)
    //
    // However the writer of this code finds it clearer to use instead:
    //   &pkt->data
    uint8_t data;
};

class RFLink;

class PktKeeper {
    private:
        Packet *pkt;

    public:
        PktKeeper();
        PktKeeper(byte pkt_len);
        ~PktKeeper();

        void release_data();

        void copy_packet(const PktKeeper* pktkeeper);
        bool check_rcvd_pkt_is_ok(const RFLink *link, byte nb_bytes);

        void prepare_for_sending(const RFLink *link, Header* header,
                                 const void *data);

        const Header* get_header_ptr() const;
        byte get_flags();
        void set_flags(byte arg_flags);

        const Packet* get_pkt_ptr_ro() const;
        // Yes, getting rid of the below would be cleaner, but to the expense of
        // some dynamic memory (used as intermediate buffer), or, more
        // complexity (ask PktKeeper class to manage reception).
        Packet* notrecommended_get_pkt_ptr();

        byte get_pkt_len() const;
        const void* get_data_ptr() const;
        byte get_data_len() const;

        void copy_data(void *buf, byte buf_len, byte* rec_len) const;
        void reduce_packet_to_its_header();
};

typedef enum {
    OPT_ADDRESS = 0,
    OPT_SNIF_MODE,
    OPT_EMISSION_POWER
} opt_t;

typedef struct {
    bool used;
    address_t src;
    mtime_t mtime;
    pktid_t last_pktid_seen;
} cache_pktid_t;

enum {
    ST_NOTHING = 0,
    ST_SEND,
    ST_SEND_DONE,
    ST_RECEIVE,
    ST_RECEIVE_DATA_AVAILABLE,
    ST_RECEIVE_DATA_RETRIEVED,
    ST_RECEIVE_TIMEDOUT,
    ST_FINISHED,
    ST_LAST
};

#define T_NONE      0
#define T_EVWAKEUP  (1 << 0)
#define T_EVPKTRCVD (1 << 1)

class Task {
    friend class RFLink;

    private:
        Task* next;

        taskid_t taskid;
        byte status;

        PktKeeper pktkeeper;

        mtime_t mtime_ref;

        mtime_t mtime_wakeup;
        byte last_retcode;

        byte nb_send_schedules;
        const mtime_t* send_schedule_ptr;
        byte send_schedule_pos;

        unsigned char evtsub_wakeup    :1;
        unsigned char evtsub_pktrcvd   :1;

        unsigned char is_an_ack        :1;
        unsigned char need_ack         :1;
        unsigned char has_received_ack :1;
        unsigned char unattended       :1;

        unsigned char rcv_from_1sender :1;

        address_t addr_rcv_from_1sender;

        byte nbsend;
};

struct RXConfig {
    unsigned char def_sender     :1;
    unsigned char def_timeout    :1;
    unsigned char def_rxcallback :1;
    address_t sender;
    mtime_t timeout;
    void (*rxcallback)(byte res, void* buf, byte buf_len, byte* rec_len);
};

struct RFLinkFunctions {
    void (*deviceInit)(byte *max_data_len, bool reset_only);
    byte (*deviceSend)(const void* data, byte len);
    byte (*deviceReceive)(void* buf, byte buf_len);
    void (*deviceSetOpt)(opt_t opt, void* data, byte len);

    void (*setInterrupt)(void (*func)());
    void (*resetInterrupt)();

    RFLinkFunctions();
};

class RFLink {
    private:

// Variables

        byte max_payload_len;

        // The below is NOT about Arduino' instructions noInterrupts() and
        // interrupts().
        // It is all about the RF device' interrupt being armed, or not.

        // Are interrupts armed?
        //   true:  interrupts are armed
        //          = execution of attachInterrupt()
        //   false: interrupts are not armed
        //          = execution of detachInterrupt()
        //          = also default situation (if none of attachInterrupt() and
        //            detachInterrupt() got called)
        unsigned char interrupt_is_attached :1;

        unsigned char device_addr_has_been_defined :1;
        unsigned char pre_allocate :1;

        unsigned char auto_sleep :1;

        address_t device_addr;

        pktid_t last_pktid;
        taskid_t last_taskid;

        mtime_t receive_data_avail_delay;
        mtime_t receive_purge_delay;
        mtime_t send_purge_delay;

        mtime_t last_device_reset;

        PktKeeper *recpkt;

        byte task_count;
        byte max_task_count;

        // Will gracefully manage packet ids (that is, discard a given packet if
        // id already seen for a given source), up to as many different sources.
        cache_pktid_t cache_pktids[PKTID_CACHE_SIZE];

#ifndef ENFORCE_MAX_TASK_COUNT_AT_COMPILE_TIME
        Task* tskhead;
#else
        Task tskhead[DEFAULT_MAX_TASK_COUNT];
#endif

        RFLinkFunctions funcs;

// Member-functions

        // "Arm" device interruptions
        void interrupts_on();
        // "Unarm" device interruptions
        void interrupts_off();

        void task_destroy(Task* tsk);
        void task_initialize(Task* tsk);
        Task* task_create(byte status);

        bool check_pktid_already_seen(address_t src, pktid_t pktid);

        Task* get_task_by_taskid(taskid_t taskid);

        void initialize_recpkt_if_necessary();

    public:

        RFLink(byte maxtask = DEFAULT_MAX_TASK_COUNT,
               unsigned char prealloc = DEFAULT_PRE_ALLOCATE);
        ~RFLink();

        void register_funcs(const RFLinkFunctions* arg_funcs);
        static byte get_header_len();

        byte get_max_payload_len() const;
        byte get_pkt_max_size() const;

        const char* get_err_string(byte errcode) const;

        void set_opt(opt_t opt, void* data, byte len);
        void set_opt_byte(opt_t opt, byte value);

        void set_auto_sleep(bool v);

        void do_events();

        byte send_noblock(taskid_t* taskid, address_t dst,
                          const void* data, byte len, bool ack);
        byte send_ack_noblock(taskid_t* taskid, Header* h);
        byte send_get_final_status(taskid_t taskid, byte *nbsend = nullptr);
        void send_ack(Task* tsk);
        byte send(address_t dst, const void* data, byte len, bool ack,
                  byte *nbsend = nullptr);

        byte tev_wakeup(Task* tsk);
        byte tev_received(Task* tsk, PktKeeper* pk, bool pktid_already_seen,
                          bool* pkt_consumed);

        byte receive_noblock(taskid_t* taskid, RXConfig* cfg = nullptr);
        byte data_retrieve(Task* tsk, void* buf, byte buf_len, byte* rec_len,
                           address_t* sender);
        byte receive(void* buf, byte buf_len, byte* rec_len,
                     address_t* sender = nullptr, RXConfig* cfg = nullptr);

        void data_retrieved_post(Task* tsk);
        byte task_get_status(taskid_t taskid);

        void delay_ms(long int d);

#ifdef RFLINK_DEBUG
        void dbg_print_status(bool is_eligible_for_sleep);
#endif

};

#endif // _RFLINK_H

